### PR TITLE
Remove BoundPods from Kubelet

### PR DESCRIPTION
**Pull request by wojtek-t**
This PR is changing Kubelet to use Pods instead of BoundPods.

This is partially addressing #5392.

cc @bgrant0607 @vmarmol @dchen1107 @erictune

### CODE REVIEW

File: pkg/api/types.go [hunk 1]
```diff
@@ -1277,7 +1277,7 @@ type BoundPods struct {
 	Host string `json:"host"`
 
 	// Items is the list of all pods bound to a given host.
-	Items []BoundPod `json:"items"`
+	Items []Pod `json:"items"`
```
Comment (vmarmol) [inline pkg/api/types.go]: Why this change? Maybe I'm missing where it is used.

**Comment by vmarmol** [comment]
This is looking good @wojtek-t, what do you think is missing? After this PR we can probably remove BoundPods entirely since they're only used here and you're removing 90% of its uses.

**Comment by erictune** [comment]
Thanks, this was really quick!

So, does this still accept data from file and http source which is in `boundPod` format?  It should.

A subsequent PR can make it so that we autodetect whether a file/http source is `containerManifest`, `boundPod` or `pod`.

**Comment by dchen1107** [comment]
LGTM. 

@erictune Why we still want boundPod format from file and http source?

**Comment by erictune** [comment]
For backward compatibility of any containerVM users who are using this format?

**Comment by dchen1107** [comment]
We never advertise / publish that. ContainerVM users should either on containerManifest or Pod, but I am not sure other channels.

File: pkg/api/conversion.go [hunk 2]
```diff
@@ -58,7 +58,7 @@ func init() {
```
Comment (yujuhong) [inline pkg/api/conversion.go]: `out.SelfLink = "/api/v1beta1/boundPods/" + in.ID` needs to be changed too.

File: pkg/api/types.go (hunk 1 above)
Comment (yujuhong) [inline pkg/api/types.go]: Agreed. Should not need to change this.

Comment (wojtek-t) [inline pkg/api/types.go]: This is something that is missing in this PR :)
Basically, in this proof of concept I'm still using "BoundPods" which I should change to "Pods".
Will change this on Monday.

**Comment by wojtek-t** [comment]
@erictune: I don't - I didn't check it yet.
BTW - do I understand correctly that in file channel we only accept "ContainerMainfest"?

**Comment by erictune** [comment]
We definitely want to accept `pod` as well as `containerManifest`.
Dawn says we can stop accepting `boundPod`.

File: pkg/api/types.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/api/types.go]: Done.

File: pkg/api/conversion.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/api/conversion.go]: Done.

**Comment by wojtek-t** [comment]
Thanks for the feedback! PTAL

I updated this PR and I think it's ready for review.
I've run e2e tests and the result is:
[Fail] Shell [It] tests that guestbook.sh passes 
Ran 25 of 27 Specs in 1027.552 seconds
FAIL! -- 24 Passed | 1 Failed | 2 Pending | 0 Skipped F0316 11:54:55.527483    8717 driver.go:87] At least one test failed
However, I think it's not related to my change - it might be related to #5431
I will check if #5452 will fix this problem and update this thread.

**Comment by wojtek-t** [comment]
BTW - I'm going to completely remove BoundPod type in the subsequent PR.

File: pkg/api/v1beta1/types.go [hunk 3]
```diff
@@ -1079,7 +1079,7 @@ type BoundPods struct {
 	Host string `json:"host" description:"name of a node that these pods were bound to"`
 
 	// Items is the list of all pods bound to a given host.
-	Items []BoundPod `json:"items" description:"list of all pods bound to a given host"`
+	Items []Pod `json:"items" description:"list of all pods bound to a given host"`
```
Comment (erictune) [inline pkg/api/v1beta1/types.go]: I'm not sure if it is okay to redefine v1beta1.
Do any apiserver apis use this type?

Comment (erictune) [inline pkg/api/v1beta1/types.go]: I guess no endpoints use it, so this is ok.

File: pkg/kubelet/config/config.go [hunk 4]
```diff
@@ -298,12 +298,12 @@ func (s *podStorage) seenSources(sources ...string) bool {
...
 }
 
-func filterInvalidPods(pods []api.BoundPod, source string, recorder record.EventRecorder) (filtered []*api.BoundPod) {
+func filterInvalidPods(pods []api.Pod, source string, recorder record.EventRecorder) (filtered []*api.Pod) {
 	names := util.StringSet{}
 	for i := range pods {
 		pod := &pods[i]
 		var errlist []error
-		if errs := validation.ValidateBoundPod(pod); len(errs) != 0 {
+		if errs := validation.ValidatePod(pod); len(errs) != 0 {
```
Comment (erictune) [inline pkg/kubelet/config/config.go]: Are there are any checks that are being added here as a result of this change, which might cause problems for e.g. ContainerVM?

File: pkg/kubelet/server.go [hunk 5]
```diff
@@ -260,12 +260,12 @@ func (s *Server) handleContainerLogs(w http.ResponseWriter, req *http.Request) {
 
 // handleBoundPods returns a list of pod bound to the Kubelet and their spec
 func (s *Server) handleBoundPods(w http.ResponseWriter, req *http.Request) {
-	pods, err := s.host.GetBoundPods()
+	pods, err := s.host.GetPods()
 	if err != nil {
 		s.error(w, err)
 		return
 	}
-	boundPods := &api.BoundPods{
```
Comment (erictune) [inline pkg/kubelet/server.go]: Are any kubelet clients going to be confused to now see a different Kind?

File: pkg/api/v1beta1/types.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/api/v1beta1/types.go]: Yes - in fact in #5502 I'm removing BoundPods struct completely.
BTW - I can revert this change if you prefer (this remains from the initial version, but currently I migrated all the usage of BoundPods to PodList).

File: pkg/kubelet/config/config.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/kubelet/config/config.go]: ValidatePod (at least currently) is validating only:
- "meta" field (which I believe is fine)
- ValidatePodSpec - this call is exactly the same inside both ValidatePod and ValidateBoundPod
So I think this change is fine.

File: pkg/kubelet/server.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/kubelet/server.go]: I can't see any client using this handler. But I only did it by code "grep" and I'm not sure how should I check it in more legitimate way...

File: pkg/api/types.go [hunk 6]
```diff
@@ -616,6 +616,21 @@ type Pod struct {
 	Status PodStatus `json:"status,omitempty"`
 }
 
+// Pods is a list of Pods bound to a common server. The resource version of
+// the pod list is guaranteed to only change when the list of pods changes.
+/*type Pods struct {
```
Comment (vmarmol) [inline pkg/api/types.go]: Commented out?

File: pkg/kubelet/server.go (hunk 5 above)
Comment (vmarmol) [inline pkg/kubelet/server.go]: @erictune's worry is about external clients. AFAIK the Kubelet API is not meant for public consumption so this change is probably okay. To be extra safe we can change the endpoint name /boundPods -> /pods.

Comment (wojtek-t) [inline pkg/kubelet/server.go]: I see - renamed /boundPods to /pods

File: pkg/api/types.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/api/types.go]: Remove - initially I didn't found PodList type, which was what I actually needed.

**Comment by wojtek-t** [comment]
PTAL

**Comment by vmarmol** [comment]
LGTM, but will wait on @erictune's LGTM as well :)

File: pkg/api/conversion.go [hunk 7]
```diff
@@ -53,12 +49,12 @@ func init() {
 			out.UID = in.UUID
 
 			if in.ID != "" {
-				out.SelfLink = "/api/v1beta1/boundPods/" + in.ID
```
Comment (vmarmol) [inline pkg/api/conversion.go]: Hmm actually these shouldn't change. Bad search replace?

Comment (wojtek-t) [inline pkg/api/conversion.go]: Why this shouldn't change? [I thought it should be changed]

Comment (vmarmol) [inline pkg/api/conversion.go]: This is an apiserver endpoint, the other is a Kubelet endpoint.

Comment (wojtek-t) [inline pkg/api/conversion.go]: I though that it's an apiserver endpoint, but still - I didn't found /boundPods handler in apiserver.
Should I revert this change?

Comment (vmarmol) [inline pkg/api/conversion.go]: I asked @yujuhong and she verified it should be "pods", sorry for the churn.

Comment (wojtek-t) [inline pkg/api/conversion.go]: Thanks for verifying!

**Comment by wojtek-t** [comment]
So IIUC, we're waiting for @erictune to LGTM, right?

**Comment by vmarmol** [comment]
Yep!

**Comment by erictune** [comment]
LGTM

**Comment by wojtek-t** [comment]
Thanks Eric - I'm squashing the commits now.

**Comment by wojtek-t** [comment]
Commits squashed - all tests are passing.

**Comment by vmarmol** [comment]
Merging, thanks @wojtek-t!

**Pull request merged**